src/libstd/sys/sgx/unwind.rs

```diff
@@ -0,0 +1,155 @@
+//! when libunwind happens to be linked.
+
+#[cfg(not(test))]
+use crate::{
+    alloc::{self, Layout},
+    lock_api::RawRwLock as _,
+    slice, str,
+    sync::atomic::Ordering,
+};
+use crate::{parking_lot::RawRwLock, sync::atomic::AtomicBool};
+
+#[cfg(not(test))]
+const EINVAL: i32 = 22;
+
+#[repr(C)]
+pub struct RwLock {
+    lock: RawRwLock,
+    is_write_locked: AtomicBool,
+}
+
+// used by libunwind port
+#[cfg(not(test))]
+#[no_mangle]
+pub unsafe extern "C" fn __rust_rwlock_rdlock(p: *mut RwLock) -> i32 {
+    if p.is_null() {
+        return EINVAL;
+    }
+    (*p).lock.lock_shared();
+    return 0;
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub unsafe extern "C" fn __rust_rwlock_wrlock(p: *mut RwLock) -> i32 {
+    if p.is_null() {
+        return EINVAL;
+    }
+    (*p).lock.lock_exclusive();
+    (*p).is_write_locked.store(true, Ordering::Relaxed);
+    return 0;
+}
+#[cfg(not(test))]
+#[no_mangle]
+pub unsafe extern "C" fn __rust_rwlock_unlock(p: *mut RwLock) -> i32 {
+    if p.is_null() {
+        return EINVAL;
+    }
+    if (*p)
+        .is_write_locked
+        .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
+        .is_ok()
+    {
+        (*p).lock.unlock_exclusive()
+    } else {
+        (*p).lock.unlock_shared();
+    }
+    return 0;
+}
+
+// the following functions are also used by the libunwind port. They're
+// included here to make sure parallel codegen and LTO don't mess things up.
+#[cfg(not(test))]
+#[no_mangle]
+pub unsafe extern "C" fn __rust_print_err(m: *mut u8, s: i32) {
+    if s < 0 {
+        return;
+    }
+    let buf = slice::from_raw_parts(m as *const u8, s as _);
+    if let Ok(s) = str::from_utf8(&buf[..buf.iter().position(|&b| b == 0).unwrap_or(buf.len())]) {
+        eprint!("{}", s);
+    }
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+// NB. used by both libunwind and libpanic_abort
+pub unsafe extern "C" fn __rust_abort() {
+    crate::sys::abort_internal();
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub unsafe extern "C" fn __rust_c_alloc(size: usize, align: usize) -> *mut u8 {
+    alloc::alloc(Layout::from_size_align_unchecked(size, align))
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub unsafe extern "C" fn __rust_c_dealloc(ptr: *mut u8, size: usize, align: usize) {
+    alloc::dealloc(ptr, Layout::from_size_align_unchecked(size, align))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mem::{self, MaybeUninit};
+    use core::array::FixedSizeArray;
+
+    // Verify that the bytes of an initialized RwLock are the same as in
+    // libunwind. If they change, `src/UnwindRustSgx.h` in libunwind needs to
+    // be changed too.
+    #[test]
+    fn test_c_rwlock_initializer() {
+        /// The value of a newly initialized `RwLock`. Which happens to be
+        /// `RawRwLock::INIT` (a zeroed `usize`), a false boolean (zero)
+        /// and then padding.
+        const RWLOCK_INIT: &[u8] = &[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
```

Comment:
As mentioned in the comment just above, `src/UnwindRustSgx.h` in [libunwind](https://github.com/fortanix/llvm-project/) needs to be updated, and this PR will need to pull in the changes in `src/ci/docker/dist-various-2/Dockerfile`. @VardhanThigle can you help with this?

Reply:
I'm aware, but good you pointed it out. This won't be merged just yet, we just do a test run. But we can probably go ahead and create a PR in `libunwind` for the change. Here you go: https://github.com/fortanix/libunwind/pull/4

It's a bit of a chicken and egg problem. But the libunwind PR must be merged first, so we can update the commit id in the dockerfile in this PR. I assume that's the way to go.

Reply:
I see that `libunwind` is not a submodule of your `llvm-project` repository. The files seem copied over? I think it's best if you update your libraries as you see fit and then ping me with the new commit id :)

Reply:
cc75f6e51b7dafe1dbd2ad16d969fcd3ff446053

Reply:
The link https://github.com/fortanix/llvm-project/archive/cc75f6e51b7dafe1dbd2ad16d969fcd3ff446053.tar.gz does not work. Nor do I see any new commits pushed to `llvm-project` 🤔

Reply:
Oh I'm so sorry, this is all my mistake. I now understand your earlier comment as well. I linked to the wrong repo. The PR should've gone to https://github.com/fortanix/llvm-project/ . I'll be deleting the libunwind repo shortly.

Reply:
I see now that you actually linked to `llvm-project` at the start of this thread. It was just me who went to your organization and picked the first repo that said `libunwind` without paying attention.

The code comment says:
> ... If they change, `src/UnwindRustSgx.h` in libunwind needs to be changed too ...

And you had a repo named exactly `libunwind` with a file at the denoted path 🤷‍♂️ 😄

Reply:
a50a70f1394b2e62d6a5d2510330eb110e31dad4

Reply:
Pushed the checksum update to the [testing branch](https://github.com/faern/rust/tree/add-parking-lot-test). Will you be able to actually test these running together? So we know it works as intended.

Reply:
I'm seeing a hang in `std::io::buffered::tests::panic_in_write_doesnt_flush_in_drop` after the second thread has exited, join never returns. Backtrace excerpt:
```
#0  0x00007ffff00b33a6 in std::parking_lot::condvar::Condvar::wait_until_internal::haec9e821d58c63ec ()
#1  0x00007ffff01192ae in std::thread::park::h55b748bc253e86f5 ()
#2  0x00007ffff007cbb2 in std::sync::mpsc::oneshot::Packet$LT$T$GT$::recv::h9344b3a4c5e28c0d ()
#3  0x00007ffff000dfce in std::sync::mpsc::Receiver$LT$T$GT$::recv::h18960999ebdd9329 ()
#4  0x00007ffff0119c9e in std::thread::JoinHandle$LT$T$GT$::join::h5220ab83fba0a8b1 ()
#5  0x00007ffff00955ee in std::io::buffered::tests::panic_in_write_doesnt_flush_in_drop::hb734fd606fd827e8 ()
```
I'm still trying to figure out if this is a bug in unwinding or in parking_lot's SGX implementation.

Reply:
This backtrace, and your description, sounds very similar to the hangs we saw when discovering the problem around `fork()` and `parking_lot` not being fork safe. See these threads: https://github.com/rust-lang/rust/pull/56410#discussion_r261896206 and https://github.com/rust-lang/rust/pull/56410#discussion_r261896521

I have no idea if this is remotely related. But just dropping this info here for now.

Reply:
Your backtrace shows the still alive thread and that it's waiting for the spawned thread to return its exiting result. I don't think the bug is in any code running in this thread. I think the interesting part is how the other thread exits and what stops it from storing/sending its exit value before dying. Could as you say very well be a bug in the unwinding code for example.

Reply:
I don't really get how it's supposed to work. I'm looking at this code: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/sgx/thread.rs#L32-L33

Is this where it executes the closure that is the thread's closure? It just runs the closure, without catching any panics. So if it unwinds, will it ever reach `let _ = self.done.send(());` and actually tell the parent thread that it's done?

Reply:
Ok, the second thread does call `notify_one_slow`, but it never calls `send`. Instead, it just calls `Instant::now` *a lot* (this is extremely slow on SGX), and then exits.

Reply:
I can find the code path `notify_one_slow -> unpark_requeue -> FairTimeout::should_timeout -> Instant::now()`. But that's the only one I can see. And none of this is in any loop, so it should only call `Instant::now()` once per `notify_one_slow` 🤔. So I have no idea how you end up in there *a lot*.

What you can try is removing all the code in `should_timeout` and replace it with a hardcoded boolean. To see if that makes any difference.

Reply:
It's also in `FairTimeout::new`
```
#0  _$LT$T$u20$as$u20$std..sys..sgx..abi..usercalls..raw..ReturnValue$GT$::from_registers::h5ba35fe8f072f078 (call=..., usercall_retval=...)
    at src/libstd/sys/sgx/abi/usercalls/raw.rs:148
#1  std::sys::sgx::abi::usercalls::raw::insecure_time::h4c93ef2825976fbe () at src/libstd/sys/sgx/abi/usercalls/raw.rs:237
#2  std::sys::sgx::abi::usercalls::insecure_time::h952732632fe35542 () at src/libstd/sys/sgx/abi/usercalls/mod.rs:141
#3  std::sys::sgx::time::Instant::now::hf22da832be6ac7c7 () at src/libstd/sys/sgx/time.rs:14
#4  std::time::Instant::now::hb5cf7a37a7f4042a () at src/libstd/time.rs:155
#5  0x00007ffff01e8d75 in std::parking_lot_core::parking_lot::FairTimeout::new::hd500c4d69f6ab960 () at src/libstd/../parking_lot/core/src/parking_lot.rs:100
#6  std::parking_lot_core::parking_lot::Bucket::new::heb1b2931ac456124 () at src/libstd/../parking_lot/core/src/parking_lot.rs:74
#7  _$LT$std..parking_lot_core..parking_lot..Bucket$u20$as$u20$core..clone..Clone$GT$::clone::hb146659cdef4b667 (self=<optimized out>)
    at src/libstd/../parking_lot/core/src/parking_lot.rs:83
#8  _$LT$alloc..vec..ExtendElement$LT$T$GT$$u20$as$u20$alloc..vec..ExtendWith$LT$T$GT$$GT$::next::hbfef038f11319fec (self=<optimized out>)
    at /home/jethro/projects/rust/src/liballoc/vec.rs:1407
#9  alloc::vec::Vec$LT$T$GT$::extend_with::he598870e4af22421 (self=<optimized out>, n=16, value=...) at /home/jethro/projects/rust/src/liballoc/vec.rs:1437
#10 0x00007ffff01fad20 in _$LT$T$u20$as$u20$alloc..vec..SpecFromElem$GT$::from_elem::h92d5c81e0940f5b1 (elem=..., n=16) at /home/jethro/projects/rust/src/liballoc/vec.rs:1546
#11 alloc::vec::from_elem::h5f517c218e474b25 (elem=<error reading variable: access outside bounds of object referenced via synthetic pointer>, n=16)
    at /home/jethro/projects/rust/src/liballoc/vec.rs:1535
#12 std::parking_lot_core::parking_lot::HashTable::new::h8f572e9bf0f18aa5 (num_threads=3, prev=0x0) at src/libstd/../parking_lot/core/src/parking_lot.rs:47
#13 std::parking_lot_core::parking_lot::create_hashtable::h500e6af0dc69c42a () at src/libstd/../parking_lot/core/src/parking_lot.rs:210
#14 0x00007ffff01cbd74 in std::parking_lot_core::parking_lot::get_hashtable::h8f96a5a990cbfe68 () at src/libstd/../parking_lot/core/src/parking_lot.rs:200
#15 std::parking_lot_core::parking_lot::lock_bucket_pair::ha45f9718df8665fe (key1=<optimized out>, key2=<optimized out>) at src/libstd/../parking_lot/core/src/parking_lot.rs:387
#16 0x00007ffff01cbf49 in std::parking_lot_core::parking_lot::unpark_requeue::he0d02ddf9a427906 (key_from=140737223877744, key_to=140737223877752, validate=..., callback=...)
    at src/libstd/../parking_lot/core/src/parking_lot.rs:857
#17 std::parking_lot::condvar::Condvar::notify_one_slow::hcb597d05bd0ade1d (self=0x7ffff03c6470, mutex=0x7ffff03c6478) at src/libstd/../parking_lot/src/condvar.rs:190
#18 0x00007ffff01ca021 in std::sync::mpsc::blocking::SignalToken::signal::h78695a323f8b43ac (self=<optimized out>) at src/libstd/sync/mpsc/blocking.rs:48
#19 std::sync::mpsc::oneshot::Packet$LT$T$GT$::send::h06585b92f34d8266 (self=<optimized out>, t=<optimized out>) at src/libstd/sync/mpsc/oneshot.rs:119
#20 0x00007ffff01d9c48 in std::sync::mpsc::Sender$LT$T$GT$::send::h480f711e8406c989 (self=0x7ffff3c9cb10, t=<optimized out>) at src/libstd/sync/mpsc/mod.rs:822
#21 0x00007ffff01d8444 in std::sys::sgx::thread::task_queue::Task::run::h121325d657af04c9 (self=...) at src/libstd/sys/sgx/thread.rs:33
```

Reply:
This is in `HashTable::new`:
```
entries: vec![Bucket::new(); new_size].into_boxed_slice(),
```
new_size is 16.

Reply:
`Bucket::new()` will call `FairTimeout::new()` which will in turn call `Instant::new()` also. But this is also not inside any loop. New `Bucket`s should only be created when the `HashTable` is created or resized. Which only happens when the number of threads increase.

If your `new_size` is 16 that should mean it calls `Instant::now()` 16 times. What order of magnitude slow are we talking here, regarding `insecure_time`? All those buckets could potentially share the same starting instant. Only create one instant and pass it down through `Bucket::new()` to `FairTimeout::new()`. Or `FairTimeout` could start out without an `Instant` and create one on first use.

Reply:
> What order of magnitude slow are we talking here, regarding insecure_time?

```
println!("{:?}", Instant::now().elapsed());
```

Linux: ~100ns
SGX: ~6µs

Reply:
Ok. ~6 times slower. But that does not explain the problem we are seeing here, does it? Why would a slightly slower hashmap resizing cause the panicking thread in the test to not finish gracefully at all?

Reply:
You could try what I now pushed to the testing branch. Both calls to `Instant::new()` should be gone from SGX, and it will always return `false` from `should_timeout`.

Reply:
Ok after some more debugging, I think the problem is similar to https://github.com/rust-lang/rust/pull/58399.

> When compiling the unit tests of std, there are basically two copies of std: one normal one, and one with cfg(test). This can lead to problems if both copies think they uniquely define a global resource.

It looks like `parking_lot` has many static variables. These need to be the same in both copies of `std`. I actually had the same problem with the task queue in SGX, see https://github.com/rust-lang/rust/pull/59136/files#diff-6346734096cfe60a79cd965a5355f629

Reply:
I assume the relevant statics in `parking_lot` would be `NUM_THREADS` and `HASHTABLE` in `core/parking_lot.rs`. They are the globals that all locks interact with.

Btw, I pinged you on IRC. If that medium is available to you it's probably a lot more effective means of communication for discussing this.